src/lib.rs

```diff
@@ -102,60 +105,120 @@ struct Reaper {
+            async_process_thread: Once::new(),
+            child_count: AtomicUsize::new(0),
+        })
+    }
+
+    /// Ensure that the reaper is driven.
+    ///
+    /// If there are no active `driver()` callers, this will spawn the `async-process` thread.
+    #[inline]
+    fn ensure_driven(&'static self) {
+        if self.drivers.load(Ordering::Acquire) == 0 {
+            self.start_driver_thread();
+        }
+    }
+
+    /// Start the `async-process` thread.
+    #[cold]
+    fn start_driver_thread(&'static self) {
+        self.async_process_thread.call_once(move || {
             thread::Builder::new()
                 .name("async-process".to_string())
-                .spawn(|| REAPER.wait_blocking().reap())
+                .spawn(move || {
+                    let driver = async move {
+                        self.drivers.fetch_add(1, Ordering::SeqCst);
+                        let guard = self.driver_guard.lock().await;
+                        self.reap(guard).await
+                    };
+
+                    #[cfg(unix)]
+                    async_io::block_on(driver);
+
+                    #[cfg(not(unix))]
+                    future::block_on(driver);
+                })
                 .expect("cannot spawn async-process thread");
-
-            Reaper {
-                sigchld: Event::new(),
-                zombies: Mutex::new(Vec::new()),
-                pipe: Pipe::new().expect("cannot create SIGCHLD pipe"),
-            }
-        })
+        });
     }
 
     /// Reap zombie processes forever.
-    fn reap(&'static self) -> ! {
+    async fn reap(&'static self, _driver_guard: async_lock::MutexGuard<'_, ()>) -> ! {
         loop {
             // Wait for the next SIGCHLD signal.
-            self.pipe.wait();
+            self.pipe.wait().await;
 
             // Notify all listeners waiting on the SIGCHLD event.
             self.sigchld.notify(std::usize::MAX);
 
-            // Reap zombie processes.
-            let mut zombies = self.zombies.lock().unwrap();
+            // Reap zombie processes, but make sure we don't hold onto the lock for too long!
+            let mut zombies = mem::take(&mut *self.zombies.lock().unwrap());
             let mut i = 0;
-            while i < zombies.len() {
-                if let Ok(None) = zombies[i].try_wait() {
-                    i += 1;
-                } else {
-                    zombies.swap_remove(i);
+            'reap_zombies: loop {
+                for _ in 0..50 {
+                    if i >= zombies.len() {
+                        break 'reap_zombies;
+                    }
+
+                    if let Ok(None) = zombies[i].try_wait() {
+                        i += 1;
+                    } else {
+                        zombies.swap_remove(i);
+                    }
                 }
+
+                // Be a good citizen; yield if there are a lot of processes.
+                future::yield_now().await;
```

Comment:
shouldn't we put the zombie processes back and unlocking before yielding?

Reply:
Since the only thing other users of the zombie list do is add to it, we don't actually have to put the zombie list back. However it is a good idea to read from it.